jobsdb/unionQuery.go

```diff
@@ -0,0 +1,284 @@
+		// event_count default 1, number of items in payload
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				`FROM %[1]s AS jobs `+
+				`LEFT JOIN %[2]s AS job_status ON jobs.job_id=job_status.job_id `+
+				`WHERE job_status.job_id is NULL AND customer='%[3]s'`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	} else {
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				` FROM %[1]s AS jobs `+
+				`WHERE customer='%[3]s' AND jobs.job_id NOT IN (SELECT DISTINCT(job_status.job_id) FROM %[2]s AS job_status)`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	}
+
+	if len(customValFilters) > 0 && !params.IgnoreCustomValFiltersInQuery {
+		sqlStatement += " AND " + constructQuery(jd, "jobs.custom_val", customValFilters, "OR")
+	}
+
+	if len(parameterFilters) > 0 {
+		sqlStatement += " AND " + constructParameterJSONQuery("jobs", parameterFilters)
+	}
+
+	//avoinding AfterJobID for now
+
+	if params.UseTimeFilter {
+		sqlStatement += fmt.Sprintf(" AND created_at < %s", params.Before)
+	}
+
+	if order {
+		sqlStatement += " ORDER BY jobs.job_id"
+	}
+
+	if count > 0 {
+		sqlStatement += fmt.Sprintf(" LIMIT %d", count)
+	}
+
+	if params.EventCount > 0 {
+		sqlStatement = fmt.Sprintf(`SELECT * FROM (`+sqlStatement+`) AS subquery WHERE running_event_counts - event_count + 1 <= %d;`, params.EventCount)
+	}
+
+	return sqlStatement
+}
+
+func (jd *HandleT) GetUnprocessedUnion(unionMap map[string]int, params GetQueryParamsT) []*JobT {
+	// if params.JobCount == 0 { //how do we use this check..? is it required anymore
+	// 	return []*JobT{}
+	// }
+
+	//add stats
+
+	//The order of lock is very important. The migrateDSLoop
+	//takes lock in this order so reversing this will cause
+	//deadlocks
+	jd.dsMigrationLock.RLock()
+	jd.dsListLock.RLock()
+	defer jd.dsMigrationLock.RUnlock()
+	defer jd.dsListLock.RUnlock()
+
+	dsList := jd.getDSList(false)
+	outJobs := make([]*JobT, 0)
+
+	//removed count assert, because that params.count is not used..?
+
+	for _, ds := range dsList {
+		jobs := jd.getUnprocessedUnionDS(ds, unionMap, params)
+		outJobs = append(outJobs, jobs...)
+		// count -= len(jobs) //this is not necessary anymore right?
+
+		//we could further limit jobs by eventCount - ignoring that for now
+		// just focusing on getting so many rows of jobs for now
+	}
+
+	return outJobs
+}
+
+func (jd *HandleT) getUnprocessedUnionDS(ds dataSetT, unionMap map[string]int, params GetQueryParamsT) []*JobT {
+	queryString := jd.getUnprocessedUnionQuerystring(unionMap, ds, params)
+
+	var rows *sql.Rows
+	var err error
+
+	rows, err = jd.dbHandle.Query(queryString)
+	jd.assertError(err)
+
+	defer rows.Close()
+
+	var jobList []*JobT
+	for rows.Next() {
+		var job JobT
+		var _null int
+		err := rows.Scan(&job.JobID, &job.UUID, &job.UserID, &job.Parameters, &job.CustomVal,
+			&job.EventPayload, &job.EventCount, &job.CreatedAt, &job.ExpireAt, &job.Customer, &_null)
+		jd.assertError(err)
+		jobList = append(jobList, &job)
+
+		unionMap[job.Customer] -= 1
```

Comment:
Shouldn't it be possible, but should we check for `job.Customer` key missing from the `unionMap`?

Reply:
the unionMap would be passed to jobsdb by the router, the keys are not updated(except deleting the key if we get all the jobs), only the values are decreased by the number of jobs we find in each DS. `job.Customer` is the same value scanned that we send in the query, so it's bound to exist in the Map. Also we're deleting the key from the map when we have retrieved the number of jobs we wanted, but the `limit` in the query helps with that situation.
Can you elaborate on more situations where we might come across this..?